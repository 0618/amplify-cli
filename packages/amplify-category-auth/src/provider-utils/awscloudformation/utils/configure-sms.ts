import { AttributeType, CognitoConfiguration } from 'amplify-cli-core';
import { CognitoStackOptions } from '../service-walkthrough-types/cognito-user-input-types';

/**
 * checks whether sms options in enabled in cognito
 */
export const configureSmsOption = (props: CognitoConfiguration | CognitoStackOptions): boolean | undefined =>
  props.autoVerifiedAttributes?.includes('phone_number') ||
  (props.mfaConfiguration !== 'OFF' && props.mfaTypes?.includes('SMS Text Message')) ||
  props.requiredAttributes?.includes('phone_number') ||
  props.usernameAttributes?.includes(AttributeType.PHONE_NUMBER);
