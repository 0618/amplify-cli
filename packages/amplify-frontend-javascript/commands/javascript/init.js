module.exports = {
    name: 'init',
    run: async (context) => {
        console.log('int js project///')
    }
}
  