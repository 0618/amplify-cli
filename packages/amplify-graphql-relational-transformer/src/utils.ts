import assert from 'assert';
import { getFieldNameFor, InvalidDirectiveError } from '@aws-amplify/graphql-transformer-core';
import { TransformerContextProvider, TransformerResourceHelperProvider } from '@aws-amplify/graphql-transformer-interfaces';
import { DirectiveNode, EnumTypeDefinitionNode, FieldDefinitionNode, Kind, ObjectTypeDefinitionNode, StringValueNode } from 'graphql';
import { getBaseType, isScalarOrEnum, toCamelCase } from 'graphql-transformer-common';
import {
  BelongsToDirectiveConfiguration,
  HasManyDirectiveConfiguration,
  HasOneDirectiveConfiguration,
  ManyToManyDirectiveConfiguration,
} from './types';

export function getRelatedTypeIndex(
  config: HasOneDirectiveConfiguration,
  ctx: TransformerContextProvider,
  indexName?: string,
): FieldDefinitionNode[] {
  const { directiveName, field, fieldNodes } = config;
  const relatedType = ctx.output.getType(config.relatedType.name.value) as any;
  const fieldMap = new Map<string, FieldDefinitionNode>();
  let partitionFieldName;
  let partitionField;
  const sortFieldNames = [];
  const sortFields = [];

  for (const field of relatedType.fields!) {
    fieldMap.set(field.name.value, field);

    for (const directive of field.directives!) {
      const directiveName = directive.name.value;
      const name = getIndexName(directive);

      if ((!indexName && directiveName === 'primaryKey') || (indexName === name && directiveName === 'index')) {
        partitionFieldName = field.name.value;

        for (const argument of directive.arguments!) {
          if (argument.name.value === 'sortKeyFields') {
            if (argument.value.kind === Kind.STRING) {
              sortFieldNames.push(argument.value.value);
            } else if (argument.value.kind === Kind.LIST) {
              for (const keyField of argument.value.values) {
                sortFieldNames.push((keyField as any).value);
              }
            }
          }
        }

        break;
      }
    }
  }

  if (partitionFieldName === undefined) {
    if (indexName) {
      throw new InvalidDirectiveError(`Index ${indexName} does not exist for model ${relatedType.name.value}`);
    }

    partitionFieldName = 'id';
  }

  partitionField = fieldMap.get(partitionFieldName);
  assert(partitionField);

  for (const sortFieldName of sortFieldNames) {
    const sortField = fieldMap.get(sortFieldName);

    assert(sortField);
    sortFields.push(sortField);
  }

  if (fieldNodes.length > 0) {
    if (getBaseType(fieldNodes[0].type) !== getBaseType(partitionField.type)) {
      throw new InvalidDirectiveError(`${fieldNodes[0].name.value} field is not of type ${getBaseType(partitionField.type)}`);
    }

    if (fieldNodes.length > 1) {
      if (sortFields.length !== fieldNodes.length - 1) {
        throw new InvalidDirectiveError(`Invalid @${directiveName} directive on ${field.name.value}. Partial sort keys are not accepted.`);
      }

      for (let i = 0; i < sortFields.length; i++) {
        const sortField = sortFields[i];
        const fieldNode = fieldNodes[i + 1];

        if (getBaseType(fieldNode.type) !== getBaseType(sortField.type)) {
          throw new InvalidDirectiveError(`${fieldNode.name.value} field is not of type ${getBaseType(sortField.type)}`);
        }
      }
    }
  }

  return [partitionField, ...sortFields];
}

export function ensureFieldsArray(config: HasManyDirectiveConfiguration | HasOneDirectiveConfiguration | BelongsToDirectiveConfiguration) {
  if (!config.fields) {
    config.fields = [];
  } else if (!Array.isArray(config.fields)) {
    config.fields = [config.fields];
  } else if (config.fields.length === 0) {
    throw new InvalidDirectiveError(`No fields passed to @${config.directiveName} directive.`);
  }
}

export function getModelDirective(objectType: ObjectTypeDefinitionNode) {
  return objectType.directives!.find(directive => {
    return directive.name.value === 'model';
  });
}

export function validateModelDirective(
  config: HasManyDirectiveConfiguration | HasOneDirectiveConfiguration | BelongsToDirectiveConfiguration | ManyToManyDirectiveConfiguration,
) {
  if (!getModelDirective(config.object)) {
    throw new InvalidDirectiveError(`@${config.directiveName} must be on an @model object type field.`);
  }
}

export function getRelatedType(
  config: HasManyDirectiveConfiguration | HasOneDirectiveConfiguration | BelongsToDirectiveConfiguration,
  ctx: TransformerContextProvider,
) {
  const { field } = config;
  const relatedTypeName = getBaseType(field.type);
  const relatedType = ctx.inputDocument.definitions.find(
    (d: any) => d.kind === Kind.OBJECT_TYPE_DEFINITION && d.name.value === relatedTypeName,
  ) as ObjectTypeDefinitionNode | undefined;

  assert(relatedType);
  return relatedType;
}

export function getFieldsNodes(
  config: HasManyDirectiveConfiguration | HasOneDirectiveConfiguration | BelongsToDirectiveConfiguration,
  ctx: TransformerContextProvider,
) {
  const { directiveName, fields, object } = config;
  const enums = ctx.output.getTypeDefinitionsOfKind(Kind.ENUM_TYPE_DEFINITION) as EnumTypeDefinitionNode[];

  return fields.map(fieldName => {
    const fieldNode = object.fields!.find(field => field.name.value === fieldName);

    if (!fieldNode) {
      throw new InvalidDirectiveError(`${fieldName} is not a field in ${object.name.value}`);
    }

    if (!isScalarOrEnum(fieldNode.type, enums)) {
      throw new InvalidDirectiveError(`All fields provided to @${directiveName} must be scalar or enum fields.`);
    }

    return fieldNode;
  });
}

export function validateRelatedModelDirective(
  config: HasManyDirectiveConfiguration | HasOneDirectiveConfiguration | BelongsToDirectiveConfiguration,
) {
  if (!getModelDirective(config.relatedType)) {
    throw new InvalidDirectiveError(`Object type ${config.relatedType.name.value} must be annotated with @model.`);
  }
}

function getIndexName(directive: DirectiveNode): string | undefined {
  for (const argument of directive.arguments!) {
    if (argument.name.value === 'name') {
      return (argument.value as StringValueNode).value;
    }
  }
}

export function getConnectionAttributeName(type: string, field: string) {
  return toCamelCase([type, field, 'id']);
}

export function getBackendConnectionAttributeName(resourceHelper: TransformerResourceHelperProvider, type: string, field: string) {
  return getConnectionAttributeName(resourceHelper.getModelNameMapping(type), field);
}

export function validateDisallowedDataStoreRelationships(
  config: HasManyDirectiveConfiguration | HasOneDirectiveConfiguration,
  ctx: TransformerContextProvider,
) {
  // If DataStore is enabled, the following scenario is not supported:
  // Model A includes a @hasOne or @hasMany relationship with Model B, while
  // Model B includes a @hasOne or @hasMany relationship back to Model A.

  if (!ctx.isProjectUsingDataStore()) {
    return;
  }

  const modelType = config.object.name.value;
  const relatedType = ctx.output.getType(config.relatedType.name.value) as ObjectTypeDefinitionNode;
  assert(relatedType);

  const hasUnsupportedConnectionFields = relatedType.fields!.some(field => {
    // If the related field has the same data type as this model, and @hasOne or @hasMany
    // is present, then the connection is unsupported.
    return (
      getBaseType(field.type) === modelType &&
      field.directives!.some(directive => {
        return directive.name.value === 'hasOne' || directive.name.value === 'hasMany';
      })
    );
  });

  if (hasUnsupportedConnectionFields) {
    throw new InvalidDirectiveError(
      `${modelType} and ${relatedType.name.value} cannot refer to each other via @hasOne or @hasMany when DataStore is in use. Use @belongsTo instead. See https://docs.amplify.aws/cli/graphql/data-modeling/#belongs-to-relationship`,
    );
  }
}

type RegisterForeignKeyMappingParams = {
  resourceHelper: TransformerResourceHelperProvider; // resourceHelper from the transformer context object
  thisTypeName: string; // the "source type" of the relation
  thisFieldName: string; // the field with the relational directive
  relatedTypeName: string; // the type that the relation points to
  relatedFieldName?: string; // the related field name that points back to this type (if any)
};

/**
 * Registers a mapping for a renamed model with a hasOne relation
 * If thisTypeName maps to a different value, it registers a mapping for the CRUD resolvers of this type
 * It also checks if the related type as been renamed and if so registers a mapping for fetching the related type through this type
 */
export function registerHasOneForeignKeyMappings({
  resourceHelper,
  thisTypeName,
  thisFieldName,
  relatedTypeName,
  relatedFieldName,
}: RegisterForeignKeyMappingParams) {
  if (resourceHelper.isModelRenamed(thisTypeName)) {
    const currAttrName = getConnectionAttributeName(thisTypeName, thisFieldName);
    const origAttrName = getBackendConnectionAttributeName(resourceHelper, thisTypeName, thisFieldName);

    (['create', 'update', 'get', 'list'] as const).forEach(op => {
      const opFieldName = getFieldNameFor(op, thisTypeName);
      const opTypeName = op === 'create' || op === 'update' ? 'Mutation' : 'Query';
      resourceHelper.addResolverFieldMapEntry(opTypeName, opFieldName, thisTypeName, [currAttrName, origAttrName], op === 'list');
    });
  }

  if (resourceHelper.isModelRenamed(relatedTypeName) && relatedFieldName !== undefined) {
    const relatedCurrAttrName = getConnectionAttributeName(relatedTypeName, relatedFieldName);
    const relatedOrigAttrName = getBackendConnectionAttributeName(resourceHelper, relatedTypeName, relatedFieldName);
    resourceHelper.addResolverFieldMapEntry(thisTypeName, thisFieldName, thisTypeName, [relatedCurrAttrName, relatedOrigAttrName], false);
  }
}

/**
 * This function is similar to registerHasOneForeignKeyMappings but subtly different
 * Because hasMany creates a foreign key field on the related type, this function registers the mapping on the related type.
 * It attaches a mapping to it's own field resolver to map the related type results as well
 */
export function registerHasManyForeignKeyMappings({
  resourceHelper,
  thisTypeName,
  thisFieldName,
  relatedTypeName,
}: RegisterForeignKeyMappingParams) {
  if (!resourceHelper.isModelRenamed(thisTypeName)) {
    return;
  }

  const currAttrName = getConnectionAttributeName(thisTypeName, thisFieldName);
  const origAttrName = getBackendConnectionAttributeName(resourceHelper, thisTypeName, thisFieldName);

  (['create', 'update', 'get', 'list'] as const).forEach(op => {
    const fieldName = getFieldNameFor(op, relatedTypeName);
    const typeName = op === 'create' || op === 'update' ? 'Mutation' : 'Query';

    // registers field mappings for CRUD resolvers on related type
    resourceHelper.addResolverFieldMapEntry(typeName, fieldName, relatedTypeName, [currAttrName, origAttrName], op === 'list');
  });

  // registers field mappings for getting the related type through the hasMany resolver
  resourceHelper.addResolverFieldMapEntry(thisTypeName, thisFieldName, relatedTypeName, [currAttrName, origAttrName], true);
}
