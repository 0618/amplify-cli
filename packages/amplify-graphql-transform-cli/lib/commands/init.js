"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const clime_1 = require("clime");
const fs = require("fs");
const log_1 = require("../log");
const TemplateSchema = `type Post @model {
    id: ID!
    title: String! @search
    tags: [String] @search
    createdAt: String
    updatedAt: String
}
`;
const DefaultConfig = (name) => `
{
    "name": "${name}"
}
`;
let default_1 = class default_1 extends clime_1.Command {
    execute() {
        /**
         * Initialize an application.
         */
        log_1.default.debug('Creating schema.graphql');
        fs.writeFileSync('schema.graphql', TemplateSchema);
        log_1.default.info(`
            Edit the schema.graphql file and then run 'appsync create schema.graphql <application-name>' to build and deploy your API.
        `);
    }
};
__decorate([
    clime_1.metadata,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], default_1.prototype, "execute", null);
default_1 = __decorate([
    clime_1.command({
        description: 'Deploy an AppSync API from your schema.graphql file',
    })
], default_1);
exports.default = default_1;
//# sourceMappingURL=init.js.map