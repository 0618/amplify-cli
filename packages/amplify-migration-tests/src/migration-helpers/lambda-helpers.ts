import {
  ExecutionContext,
  FunctionRuntimes,
  getRuntimeDisplayName,
  getTemplateChoices,
  moveUp,
  runtimeChoices,
  singleSelect,
} from '@aws-amplify/amplify-e2e-core';

export const selectRuntimePreV12 = (chain: ExecutionContext, runtime: FunctionRuntimes) => {
  const runtimeName = getRuntimeDisplayName(runtime);
  chain.wait('Choose the runtime that you want to use:');

  // reset cursor to top of list because node is default but it throws off offset calculations
  moveUp(chain, runtimeChoices.indexOf(getRuntimeDisplayName('nodejs')));

  singleSelect(chain, runtimeName, runtimeChoices);
};

export const selectTemplatePreV12 = (chain: ExecutionContext, functionTemplate: string, runtime: FunctionRuntimes) => {
  const templateChoices = getTemplateChoices(runtime);
  chain.wait('Choose the function template that you want to use');

  // reset cursor to top of list because Hello World is default but it throws off offset calculations
  moveUp(chain, templateChoices.indexOf('Hello World'));

  singleSelect(chain, functionTemplate, templateChoices);
};
