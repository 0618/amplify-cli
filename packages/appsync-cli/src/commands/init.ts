import { Command, command, param, metadata } from 'clime';
import File from '../types/File';
import GraphQLTransform from 'graphql-transform';
import SimpleTransform from 'simple-appsync-transform'
import * as fs from 'fs'
import log from '../log'

const TemplateSchema = `type Post @model {
    id: ID!
    title: String! @search
    tags: [String] @search
    createdAt: String
    updatedAt: String
}
`
const DefaultConfig = (name: string) => `
{
    "name": "${name}"
}
`

@command({
    description: 'Deploy an AppSync API from your schema.graphql file',
})
export default class extends Command {
    @metadata
    public execute() {
        /**
         * Initialize an application.
         */
        log.debug('Creating schema.graphql')
        fs.writeFileSync('schema.graphql', TemplateSchema)
        log.info(`
            Edit the schema.graphql file and then run 'appsync create schema.graphql <application-name>' to build and deploy your API.
        `)
    }
}
