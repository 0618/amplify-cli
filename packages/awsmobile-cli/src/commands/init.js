const analyzeProject = require('../lib/init-steps/s0-analyzeProject'); 
const configSettings = require('../lib/init-steps/s1-configSettings'); 
const initProviders = require('../lib/init-steps/s2-initProviders'); 
const onFailure = require('../lib/init-steps/s9-onFailure'); 
const onSuccess = require('../lib/init-steps/s9-onSuccess'); 

module.exports = {
    name: 'init',
    run: async (context) => {
        analyzeProject.run(context)
        .then(configSettings.run)
        .then(initProviders.run)
        .then(onSuccess.run)
        .catch(onFailure); 
        
    }
}
  