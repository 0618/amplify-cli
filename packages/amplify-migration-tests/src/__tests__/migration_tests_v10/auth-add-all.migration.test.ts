import {
    addAuthWithDefault,
    addAuthWithMaxOptions,
    amplifyPull,
    amplifyPushAuth,
    amplifyPushWithoutCodegen,
    createNewProjectDir,
    deleteProject,
    deleteProjectDir,
    getAppId,
    removeAuthWithDefault,
  } from '@aws-amplify/amplify-e2e-core';
  import { versionCheck, allowedVersionsToMigrateFrom } from '../../migration-helpers';
  import { initJSProjectWithProfileV10 } from '../../migration-helpers-v10/init';
import { assertNoParameterChangesBetweenProjects, collectCloudformationDiffBetweenProjects } from '../../migration-helpers/utils';
  
  describe('amplify migration test auth', () => {
    let projRoot1: string;
  
    beforeAll(async () => {
      const migrateFromVersion = { v: 'unintialized' };
      const migrateToVersion = { v: 'unintialized' };
      await versionCheck(process.cwd(), false, migrateFromVersion);
      await versionCheck(process.cwd(), true, migrateToVersion);
      console.log(`Test migration from: ${migrateFromVersion.v} to ${migrateToVersion.v}`);
      expect(migrateFromVersion.v).not.toEqual(migrateToVersion.v);
      expect(allowedVersionsToMigrateFrom).toContain(migrateFromVersion.v);
    });
  
    beforeEach(async () => {
      projRoot1 = await createNewProjectDir('authMigration1');
    });
  
    afterEach(async () => {
      // note - this deletes the original project using the latest codebase
      await deleteProject(projRoot1, null, true);
      deleteProjectDir(projRoot1);
    });
  
    it('...should add auth with max options and work on the latest version', async () => {
      await initJSProjectWithProfileV10(projRoot1, { name: 'authTest', disableAmplifyAppCreation: false });
      
      await addAuthWithMaxOptions(projRoot1, {});
      await amplifyPushAuth(projRoot1);

      const appId = getAppId(projRoot1);
      expect(appId).toBeDefined();
      const projRoot2 = await createNewProjectDir('authMigration2');
      try {
        await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
        assertNoParameterChangesBetweenProjects(projRoot1, projRoot2);
        expect(collectCloudformationDiffBetweenProjects(projRoot1, projRoot2)).toMatchSnapshot();
        await amplifyPushWithoutCodegen(projRoot2, true);
        assertNoParameterChangesBetweenProjects(projRoot1, projRoot2);
        expect(collectCloudformationDiffBetweenProjects(projRoot1, projRoot2)).toMatchSnapshot();

        // should be able to remove & add auth after pulling down an older project
        await removeAuthWithDefault(projRoot2, true);
        await addAuthWithDefault(projRoot2, {}, true);
        await amplifyPushAuth(projRoot2, true);
      } finally {
        deleteProjectDir(projRoot2);
      }
    });
  });
  