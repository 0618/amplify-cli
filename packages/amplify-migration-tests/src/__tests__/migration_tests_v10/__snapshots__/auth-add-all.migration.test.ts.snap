// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify migration test auth ...should add auth with max options and work on the latest version 1`] = `
"IAM Statement Changes
┌───┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────┬──────────────────────┬────────────────────────────┬───────────┐
│   │ Resource                                                                                                                                           │ Effect │ Action               │ Principal                  │ Condition │
├───┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼──────────────────────┼────────────────────────────┼───────────┤
│ - │ {\\"Fn::Sub\\":[\\"arn:aws:logs:\${AWS::Region}:\${AWS::AccountId}:log-group:/aws/lambda/\${lambdaName}:log-stream:*\\",{\\"lambdaName\\":\\"\${RoleMapFunction}\\"}]} │ Allow  │ logs:CreateLogGroup  │ AWS:\${LambdaExecutionRole} │           │
│   │                                                                                                                                                    │        │ logs:CreateLogStream │                            │           │
│   │                                                                                                                                                    │        │ logs:PutLogEvents    │                            │           │
└───┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────┴──────────────────────┴────────────────────────────┴───────────┘
(NOTE: There may be security-related changes not in this list. See https://github.com/aws/aws-cdk/issues/1299)

Resources
[-] AWS::IAM::Role LambdaExecutionRole destroy
[-] AWS::Lambda::Function RoleMapFunction destroy
[-] AWS::IAM::Policy LambdaCloudWatchPolicy destroy
[-] Custom::LambdaCallout RoleMapFunctionInput destroy

IAM Statement Changes
┌───┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────┬────────────────────────────────────┬───────────────────────────┬───────────┐
│   │ Resource                                                                                                                                                                                               │ Effect │ Action                             │ Principal                 │ Condition │
├───┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ \${UserPool.Arn}                                                                                                                                                                                        │ Allow  │ cognito-idp:DescribeUserPoolClient │ AWS:\${UserPoolClientRole} │           │
│ - │ \${UserPool.Arn}                                                                                                                                                                                        │ Allow  │ cognito-idp:CreateUserPoolDomain   │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                                        │        │ cognito-idp:DeleteUserPoolDomain   │                           │           │
│   │                                                                                                                                                                                                        │        │ cognito-idp:DescribeUserPool       │                           │           │
│ - │ \${UserPool.Arn}                                                                                                                                                                                        │ Allow  │ cognito-idp:CreateIdentityProvider │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                                        │        │ cognito-idp:DeleteIdentityProvider │                           │           │
│   │                                                                                                                                                                                                        │        │ cognito-idp:ListIdentityProviders  │                           │           │
│   │                                                                                                                                                                                                        │        │ cognito-idp:UpdateIdentityProvider │                           │           │
│ - │ \${UserPool.Arn}                                                                                                                                                                                        │ Allow  │ cognito-idp:UpdateUserPoolClient   │ AWS:\${UserPoolClientRole} │           │
├───┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ *                                                                                                                                                                                                      │ Allow  │ cognito-idp:DescribeUserPoolDomain │ AWS:\${UserPoolClientRole} │           │
│ - │ *                                                                                                                                                                                                      │ Allow  │ cognito-idp:DescribeUserPoolDomain │ AWS:\${UserPoolClientRole} │           │
├───┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ {\\"Fn::Sub\\":[\\"arn:aws:logs:\${region}:\${account}:log-group:/aws/lambda/\${lambda}:log-stream:*\\",{\\"region\\":\\"\${AWS::Region}\\",\\"account\\":\\"\${AWS::AccountId}\\",\\"lambda\\":\\"\${HostedUICustomResource}\\"}]}          │ Allow  │ logs:CreateLogGroup                │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                                        │        │ logs:CreateLogStream               │                           │           │
│   │                                                                                                                                                                                                        │        │ logs:PutLogEvents                  │                           │           │
├───┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ {\\"Fn::Sub\\":[\\"arn:aws:logs:\${region}:\${account}:log-group:/aws/lambda/\${lambda}:log-stream:*\\",{\\"region\\":\\"\${AWS::Region}\\",\\"account\\":\\"\${AWS::AccountId}\\",\\"lambda\\":\\"\${HostedUIProvidersCustomResource}\\"}]} │ Allow  │ logs:CreateLogGroup                │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                                        │        │ logs:CreateLogStream               │                           │           │
│   │                                                                                                                                                                                                        │        │ logs:PutLogEvents                  │                           │           │
├───┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ {\\"Fn::Sub\\":[\\"arn:aws:logs:\${region}:\${account}:log-group:/aws/lambda/\${lambda}:log-stream:*\\",{\\"region\\":\\"\${AWS::Region}\\",\\"account\\":\\"\${AWS::AccountId}\\",\\"lambda\\":\\"\${OAuthCustomResource}\\"}]}             │ Allow  │ logs:CreateLogGroup                │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                                        │        │ logs:CreateLogStream               │                           │           │
│   │                                                                                                                                                                                                        │        │ logs:PutLogEvents                  │                           │           │
├───┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ {\\"Fn::Sub\\":[\\"arn:aws:logs:\${region}:\${account}:log-group:/aws/lambda/\${lambda}:log-stream:*\\",{\\"region\\":\\"\${AWS::Region}\\",\\"account\\":\\"\${AWS::AccountId}\\",\\"lambda\\":\\"\${UserPoolClientLambda}\\"}]}            │ Allow  │ logs:CreateLogGroup                │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                                        │        │ logs:CreateLogStream               │                           │           │
│   │                                                                                                                                                                                                        │        │ logs:PutLogEvents                  │                           │           │
└───┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────┴────────────────────────────────────┴───────────────────────────┴───────────┘
(NOTE: There may be security-related changes not in this list. See https://github.com/aws/aws-cdk/issues/1299)

Parameters
[-] Parameter facebookAppId: {\\"Type\\":\\"String\\"}
[-] Parameter googleClientId: {\\"Type\\":\\"String\\"}
[-] Parameter amazonAppId: {\\"Type\\":\\"String\\"}
[-] Parameter appleAppId: {\\"Type\\":\\"String\\"}
[~] Parameter hostedUIProviderCreds: {\\"Type\\":\\"String\\"} to {\\"Type\\":\\"String\\",\\"NoEcho\\":true}

Conditions
[-] Condition ShouldOutputAppClientSecrets: {\\"Fn::Equals\\":[{\\"Ref\\":\\"userpoolClientGenerateSecret\\"},true]}

Resources
[-] AWS::Lambda::Function UserPoolClientLambda destroy
[-] AWS::IAM::Policy UserPoolClientLambdaPolicy destroy
[-] AWS::IAM::Policy UserPoolClientLogPolicy destroy
[-] Custom::LambdaCallout UserPoolClientInputs destroy
[-] AWS::Lambda::Function HostedUICustomResource destroy
[-] AWS::IAM::Policy HostedUICustomResourcePolicy destroy
[-] AWS::IAM::Policy HostedUICustomResourceLogPolicy destroy
[-] Custom::LambdaCallout HostedUICustomResourceInputs destroy
[-] AWS::Lambda::Function OAuthCustomResource destroy
[-] AWS::IAM::Policy OAuthCustomResourcePolicy destroy
[-] AWS::IAM::Policy OAuthCustomResourceLogPolicy destroy
[-] Custom::LambdaCallout OAuthCustomResourceInputs destroy
[+] AWS::Cognito::UserPoolDomain HostedUIDomainResource 
[~] AWS::Cognito::UserPoolClient UserPoolClientWeb 
 ├─ [+] AllowedOAuthFlows
 │   └─ [\\"code\\"]
 ├─ [+] AllowedOAuthFlowsUserPoolClient
 │   └─ true
 ├─ [+] AllowedOAuthScopes
 │   └─ [\\"phone\\",\\"email\\",\\"openid\\",\\"profile\\",\\"aws.cognito.signin.user.admin\\"]
 ├─ [+] CallbackURLs
 │   └─ [\\"https://signin1/\\"]
 ├─ [+] LogoutURLs
 │   └─ [\\"https://signout1/\\"]
 ├─ [+] SupportedIdentityProviders
 │   └─ [\\"Facebook\\",\\"Google\\",\\"LoginWithAmazon\\",\\"SignInWithApple\\",\\"COGNITO\\"]
 └─ [~] DependsOn
     └─ @@ -1,3 +1,4 @@
        [ ] [
        [+]   \\"HostedUIProvidersCustomResourceInputs\\",
        [ ]   \\"UserPool\\"
        [ ] ]
[~] AWS::Cognito::UserPoolClient UserPoolClient 
 ├─ [+] AllowedOAuthFlows
 │   └─ [\\"code\\"]
 ├─ [+] AllowedOAuthFlowsUserPoolClient
 │   └─ true
 ├─ [+] AllowedOAuthScopes
 │   └─ [\\"phone\\",\\"email\\",\\"openid\\",\\"profile\\",\\"aws.cognito.signin.user.admin\\"]
 ├─ [+] CallbackURLs
 │   └─ [\\"https://signin1/\\"]
 ├─ [+] LogoutURLs
 │   └─ [\\"https://signout1/\\"]
 └─ [+] SupportedIdentityProviders
     └─ [\\"Facebook\\",\\"Google\\",\\"LoginWithAmazon\\",\\"SignInWithApple\\",\\"COGNITO\\"]
[~] AWS::IAM::Role UserPoolClientRole 
 └─ [-] DependsOn
     └─ [\\"UserPoolClient\\"]
[~] AWS::Cognito::IdentityPool IdentityPool 
 └─ [-] DependsOn
     └─ [\\"UserPoolClientInputs\\"]
[~] AWS::Cognito::IdentityPoolRoleAttachment IdentityPoolRoleMap 
 ├─ [+] RoleMappings
 │   └─ {\\"UserPoolClientRoleMapping\\":{\\"AmbiguousRoleResolution\\":\\"AuthenticatedRole\\",\\"IdentityProvider\\":{\\"Fn::Sub\\":[\\"cognito-idp.\${region}.amazonaws.com/\${userPool}:\${client}\\",{\\"region\\":{\\"Ref\\":\\"AWS::Region\\"},\\"userPool\\":{\\"Ref\\":\\"UserPool\\"},\\"client\\":{\\"Ref\\":\\"UserPoolClient\\"}}]},\\"Type\\":\\"Token\\"},\\"UserPoolWebClientRoleMapping\\":{\\"AmbiguousRoleResolution\\":\\"AuthenticatedRole\\",\\"IdentityProvider\\":{\\"Fn::Sub\\":[\\"cognito-idp.\${region}.amazonaws.com/\${userPool}:\${webClient}\\",{\\"region\\":{\\"Ref\\":\\"AWS::Region\\"},\\"userPool\\":{\\"Ref\\":\\"UserPool\\"},\\"webClient\\":{\\"Ref\\":\\"UserPoolClientWeb\\"}}]},\\"Type\\":\\"Token\\"}}
 └─ [~] DependsOn
     └─ @@ -1,3 +1,5 @@
        [ ] [
        [-]   \\"IdentityPool\\"
        [+]   \\"IdentityPool\\",
        [+]   \\"UserPoolClient\\",
        [+]   \\"UserPoolClientWeb\\"
        [ ] ]

Outputs
[-] Output AppClientSecret: {\\"Value\\":{\\"Fn::GetAtt\\":[\\"UserPoolClientInputs\\",\\"appSecret\\"]},\\"Condition\\":\\"ShouldOutputAppClientSecrets\\"}
[-] Output GoogleWebClient: {\\"Value\\":{\\"Ref\\":\\"googleClientId\\"}}
[-] Output FacebookWebClient: {\\"Value\\":{\\"Ref\\":\\"facebookAppId\\"}}
[-] Output AmazonWebClient: {\\"Value\\":{\\"Ref\\":\\"amazonAppId\\"}}
[-] Output AppleWebClient: {\\"Value\\":{\\"Ref\\":\\"appleAppId\\"}}

"
`;
